[IMP] website: prevent text generation issues with foreign languages

Guillaume-gdi opened the pull request:
This commit permits to check if the text theme is translated before asking IAP for a text generation. We need the text to already be in the target language to be able to generate the text in the target language.

task-3248852

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/136937).

Guillaume-gdi commented:
Note here that it would be better if we were able to check the % of translated **words** instead of % of translated **snippets**. There are cases where a snippet is partially translated and for now we consider it translated..
I hesitate if it's overkill to try to do something or not.. We could someting like:
But maybe we could try with something like
```py

qsm-odoo commented on addons/website/models/website.py:
```diff
@@ -612,26 +612,51 @@ def _render_snippet(key):
-        except AccessError:
-            # If IAP is broken continue normally (without generating text)
-            pass
+        if text_generation_target_lang != 'en_US':
+            translated_ratio = nb_terms_translated / (nb_terms_not_translated + nb_terms_translated)
+            logger.info("Number of not translated terms: %s", nb_terms_not_translated)
+            logger.info("Number of translated terms: %s", nb_terms_translated)
+        else:
+            translated_ratio = 1
+
+        logger.info("Ratio of translated content for %s: %s", text_generation_target_lang, translated_ratio)
```
I think we don't need a difference for english or not in terms of log.
- I don't know if there is any convention but I would not log this info. Maybe use `logger.debug` instead so that you can have the log when in debug mode but not all the time
- Instead of 3 logs, only keep the ratio one, and you can put something like "55% (11/20)" (if 11 translated over 20 for example).

qsm-odoo commented on addons/website/models/website.py:
```diff
@@ -612,26 +612,51 @@ def _render_snippet(key):
+                # If IAP is broken continue normally (without generating text)
+                pass
+        else:
+            logger.warning("Not enough translated content to generate text with chat GPT")
```
Here I would not warn but log, with the same info (the ratio) as above. What do you think? As a community user, I see this warning I cannot do anything in my DB to fix that warning.

qsm-odoo commented on addons/website/models/website.py:
```diff
@@ -612,26 +612,51 @@ def _render_snippet(key):
+                    for from_lang_term, to_lang_terms in translation_dictionary.items():
+                        if from_lang_term == to_lang_terms[text_generation_target_lang]:
+                            nb_terms_not_translated += 1
+                        else:
+                            nb_terms_translated += 1
```
> I would have to have another variable like nb_total_translated_terms

That's what I meant of course, yes. Feels weird to me to count items that are X and the items that are not X instead of just counting the ones that are X while you already have in your hands the total amout of items per snippets.

\+ the use of a list comprehension instead of the loop + if + else + `+= 1` which is way more python-like to me (but again, not an expert).

But as you wish, you decide :)

Guillaume-gdi commented on addons/website/models/website.py:
```diff
@@ -608,30 +610,58 @@ def _render_snippet(key):
             placeholders = [_compute_placeholder(term) for term in terms]
 
+            if text_must_be_translated_for_openai:
+                nonlocal nb_terms_translated
+                nonlocal nb_terms_total
```
`nonlocal` is rare in the code base (about ten occurrences). We need it because we write/read a variable defined in the parent function. The alternative would be to have arrays of terms (total and translated) and take into account the size of these arrays to calculate the translation ratio.

qsm-odoo replied:
You don't need that because the current implementation seems wrong :grimacing: You don't cache those values here :grimacing: Here it means that if you render the same snippet twice, you don't increase the total terms. It could be arguably ok but it means that if you use 9 times a very poorly translated snippet, combined with 9 nicely translated other ones, you will have a coverage of ~90% while it is in fact ~50%. What you need here is including the nb translated and total in the `data` variable here (the cache) and not increase the external total in here (but outside, where you received the values as the return value of `_render_snippet`).

Guillaume-gdi replied:
If the same text appears several times, it will only appear once in `generated_content`. See => `generated_content[placeholder] = ''`.  For OpenAI to give a good result, +80% of the terms sent to it must be translated. This is different from saying that 80% of the theme's text must be translated. This nuance is important, as there is often repetition of text in themes.

> if you use 9 times a very poorly translated snippet, combined with 9 nicely translated other ones, you will have a coverage of ~90% while it is in fact ~50%. 

So in this case OpenAI will receive 90% of placeholder in the correct lang.

Maybe I can add a comment to explain that difference between theme translation ratio and placeholders translation ratio.

qsm-odoo replied:
Indeed, I was wrong :) But then all the implementation in this PR were wrong no? If you count the translated terms at each snippet rendering extraction, you may have the same text counted twice in different snippets while it will only be sent only once to IAP? Basically if a snippet contains another with a t-snippet-call, it's wrong? What about doing this then: instead of counting the terms at each snippet rendering, build an object like `generated_content` => `translated_content[from] = to`. That way if a term is repeated multiple times, you will only count it once. And counting will be very easy at the end: it will be number of from=to in `translated_content`  over the  size of `translated_content`.

Guillaume-gdi commented:
Ok @qsm-odoo, in addition to your comments, I
- removed numerical string from the ratio calculation ("8" == "8" in most lang)
- fixed an issue with en_XX lang where XX is not `US`.

qsm-odoo commented on addons/website/models/website.py:
```diff
@@ -608,30 +610,58 @@ def _render_snippet(key):
             return data
 
+        text_generation_target_lang = self.get_current_website().default_lang_id.code
+        # If the target language is not English, we need a good translation
+        # coverage. But if the target lang is en_XX it's ok to have en_US text.
+        text_must_be_translated_for_openai = not text_generation_target_lang.startswith('en_')
```
Good idea :+1:

Guillaume-gdi commented:
@qsm-odoo This is ready for your review :+1:

qsm-odoo reviewed (approved):
@robodoo r+
